k8s/reconcilers/deployment.go

```diff
@@ -0,0 +1,689 @@
+	annotations := object.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	annotations[operationStartTimeKey] = time.Now().Format(time.RFC3339)
+	object.SetAnnotations(annotations)
+}
+
+func (r *DeploymentReconciler) ensureOperationState(annotations map[string]string, objectStatus *k8smodel.DeployableStatus, provisioningState string) {
+	objectStatus.ProvisioningStatus.Status = provisioningState
+	objectStatus.ProvisioningStatus.OperationID = annotations[constants.AzureOperationIdKey]
+}
+
+func (r *DeploymentReconciler) updateObjectStatus(ctx context.Context, object Reconcilable, summaryResult *model.SummaryResult, opts patchStatusOptions, log logr.Logger) (provisioningState string, err error) {
+	status := r.determineProvisioningStatus(ctx, object, summaryResult, opts, log)
+	originalStatus := object.GetStatus()
+	nextStatus := originalStatus.DeepCopy()
+
+	r.patchBasicStatusProps(ctx, object, summaryResult, status, nextStatus, opts, log)
+	r.patchComponentStatusReport(ctx, object, summaryResult, nextStatus, log)
+	r.updateProvisioningStatus(ctx, object, summaryResult, status, nextStatus, opts, log)
+
+	if reflect.DeepEqual(&originalStatus, nextStatus) {
+		return string(status), nil
+	}
+	nextStatus.LastModified = metav1.Now()
+	object.SetStatus(*nextStatus)
+	return string(status), r.kubeClient.Status().Update(context.Background(), object)
+}
+
+func (r *DeploymentReconciler) determineProvisioningStatus(ctx context.Context, object Reconcilable, summaryResult *model.SummaryResult, opts patchStatusOptions, log logr.Logger) utilsmodel.ProvisioningStatus {
+	if opts.terminalErr != nil {
+		// add more details of the terminal error to the status
+		return utilsmodel.ProvisioningStatusFailed
+	}
+
+	if opts.nonTerminalErr != nil || summaryResult == nil || !r.hasParity(ctx, object, summaryResult, log) || opts.deploymentQueued {
+		return utilsmodel.GetNonTerminalStatus(object)
+	}
+
+	summary := summaryResult.Summary
+	switch summaryResult.State {
+	case model.SummaryStateDone:
+		// Honor OSS changes: https://github.com/eclipse-symphony/symphony/pull/148
+		// Use AllAssignedDeployed instead of targetCount/successCount to verify deployment.
+		status := utilsmodel.ProvisioningStatusSucceeded
+		if !summary.AllAssignedDeployed {
+			status = utilsmodel.ProvisioningStatusFailed
+		}
+		return status
+	default:
+		return utilsmodel.GetNonTerminalStatus(object)
+	}
+}
+
+func (r *DeploymentReconciler) patchBasicStatusProps(ctx context.Context, object Reconcilable, summaryResult *model.SummaryResult, status utilsmodel.ProvisioningStatus, objectStatus *k8smodel.DeployableStatus, opts patchStatusOptions, log logr.Logger) {
+	if objectStatus.Properties == nil {
+		objectStatus.Properties = make(map[string]string)
+	}
+	defer func() { // keeping for backward compatibility. Ideally we should remove this and use the provisioning status and provisioning status output
```

Comment:
Do we want nonTerminalErr message to overwrite all other status-details, even when terminalErr is not nil?

Reply:
I checked the code, there's no case that terminalErr and nonTerminalErr both configured. 
https://github.com/msftcoderdjw/eclipse-symphony/blob/23e42b79344cc64ea82c98315d6add6a0ad3261d/k8s/reconcilers/deployment.go#L158
https://github.com/msftcoderdjw/eclipse-symphony/blob/23e42b79344cc64ea82c98315d6add6a0ad3261d/k8s/reconcilers/deployment.go#L171
https://github.com/msftcoderdjw/eclipse-symphony/blob/23e42b79344cc64ea82c98315d6add6a0ad3261d/k8s/reconcilers/deployment.go#L263
https://github.com/msftcoderdjw/eclipse-symphony/blob/23e42b79344cc64ea82c98315d6add6a0ad3261d/k8s/reconcilers/deployment.go#L274
https://github.com/msftcoderdjw/eclipse-symphony/blob/23e42b79344cc64ea82c98315d6add6a0ad3261d/k8s/reconcilers/deployment.go#L341C1-L346C3

And I think this aligns with existing implementation. When it is terminalErr, we will set the status-details as the error message. When it is nonTerminalErr, we concatenate nonTerminalErr and status to form the status detail message.